ipq806x: 5.15: multiple fixup related to smem and others

Ansuel opened the pull request:
When 5.15 kernel was refreshed, a new implementation for smem was used, but I didn't notice that this was supported only with kernel 5.16. A backport patch should fix that.

Also backport one mtd parser patch that mute a confusing warning and revert unwanted DSA conversion for asrock G10.

Fixes: #11000

---

I would love some testing from the smem based device if this fix the bootloop... I tested this on my r7800 using smem and partition gots correctly detected with this.

Ansuel committed:
generic: 5.15: backport qcom smem patch for reserved-space support

In new kernel version from 5.16, smem node can be declared directly in
the reserved-space node. Upstream ipq806x (and to-be-merged) ipq807x
allign to this new implementation. Backport this patch to kernel 5.15 to
fix support for smem parser for ipq806x target.

Fixes: 88bf6525251f ("ipq806x: 5.15: replace dtsi patches with upstream version")
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
generic: 5.15: backport smempart parser fixup patch with EPROBE_DEFER error

Backport patch from kernel 5.15 that mute error on EPROBE_DEFER with
smempart parser.

This parser require the smem device to be probed first and currently it
may happen that mtd gets probed before the smem device causing an error
on the smempart parser. This error may be confusing and should be muted.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
ipq806x: 5.15: revert unwanted DSA conversion for ASRock G10

In refreshing DTS to the upstream version an unwanted change slipped in
the commit. The ASRock G10 dts got converted to DSA without any support.

Revert this to swconfig driver to restore normal functionality.

Fixes: 88bf6525251f ("ipq806x: 5.15: replace dtsi patches with upstream version")
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel commented:
@koerhen can you test this?

koerhen commented:
Well, not so good news. I applied the two patches to the current master (I excluded the one about ASRock) by dropping them into the 5.15 patches directory of ipq806x, and double-checked that the new patches were applied after the build in  
./build_dir/target-arm_cortex-a15+neon-vfpv4_musl_eabi/linux-ipq806x_generic/linux-5.15.74/

The bootloop still exists and I needed to unbrick the NBG6817.

ynezz commented:
> The bootloop still exists and I needed to unbrick the NBG6817.

@koerhen I don't know this platform, but I'm aware, that according to @hnyman, there is working ramoops available on R7800/XR500 devices (see 549f8694c30 and cf346dfadf5b) so maybe it would be worth trying? I assume, that your U-Boot doesn't contain [pstore](https://u-boot.readthedocs.io/en/latest/usage/cmd/pstore.html?highlight=pstore#retrieve-logs-in-u-boot) command, but IMO it should be possible to dump that ring buffer from within U-Boot by using some basic memory dump commands, which might be available there.

Ansuel commented:
Main problem is that nbg use some so it's difficult to create space for
pstore... Could be that the SDC thing is broken? It's the main difference
of ngb compared to the others... Just as a test we can try by disabling
it... System should boot by the spi anyway... In theory

Il Ven 21 Ott 2022, 10:14 Petr Štetiar ***@***.***> ha
scritto:

> The bootloop still exists and I needed to unbrick the NBG6817.
>
> @koerhen <https://github.com/koerhen> I don't know this platform, but I'm
> aware, that according to @hnyman <https://github.com/hnyman>, there is
> working ramoops available on R7800/XR500 devices (see 549f869
> <https://github.com/openwrt/openwrt/commit/549f8694c308761c25690a519de724ea3ad2bd76>
> and cf346df
> <https://github.com/openwrt/openwrt/commit/cf346dfadf5b8a40ae814b6fc171d6a4c4b64b9c>)
> so maybe it would be worth trying? I assume, that your U-Boot doesn't
> contain pstore
> <https://u-boot.readthedocs.io/en/latest/usage/cmd/pstore.html?highlight=pstore#retrieve-logs-in-u-boot>
> command, but IMO it should be possible to dump that ring buffer from within
> U-Boot by using some basic memory dump commands, which might be available
> there.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/11032#issuecomment-1286621015>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQVZRYBFHVMVKOMTX4DWEJGHXANCNFSM6AAAAAARKH5PJE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Ansuel committed:
ipq806x: 5.15: backport qcom_nandc patch for unprotected spare data fix

We currently ignore ret of the nandc partition parser if unprotected
spare data is true. This is the case for ipq806x nand.

Backport patch that fix this error and correctly handle error from
partition parser.

Fixes: ae6a63bc97cf ("ipq806x: 5.15: replace nandc patch with upstream version")
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel commented:
@koerhen ok i'm pretty confident this will now work!

koerhen commented:
Sorry for the bad news. Clean compile with the three relevant patches,  bootloop is still there and I needed to unbrick again.

koerhen commented:
@Ansuel-
sorry did not work. I applied the patches on the current master, made a clean default image (not even luci)  and the bootloop still persists. 
I needed to unbrick again.

Ansuel commented:
@pkgadd can you by chance test this on your g10 just to exclude smem problems?

Ansuel committed:
ipq806x: 5.15: drop mmc-ddr-1_8v from sdcc1 node

Zyxel NGB6817 is the only router that use mmc for rootfs. Upstream
kernel dtsi have mmc-ddr-1_8v enabled for sddc1. This is wrong as mmc on
ipq806x is supplied by a fixed 3.3v regulator and can't operate at 1.8v.
This cause the sddc1 to malfunction and cause kernel panic.
In old 5.15 version this was disabled but it was put in addition to many
other changes so it was dropped silently. Restore this patch to fix
working condition of such router.

Fixes: 88bf652 ("ipq806x: 5.15: replace dtsi patches with upstream version")
Fixes: #11000
Tested-by: Hendrik Koerner <koerhen@web.de>
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel commented:
@koerhen I'm comparing original ipq8064 dtsi and the current version and I notice this difference... wonder if it's the cause of the panic

(the panic is 99% caused by not finding rootfs and rootfs on nbg is on the ssdc1 and there we just did 2 changes this and the compatible... if they are not these 2 changes... then there is a problem with pins....)

Again thanks for the multiple test and thanks for not getting annoyed by this.

pkgadd commented:
> @pkgadd can you by chance test this on your g10 just to exclude smem problems?

Will do, as soon as I've recovered the g10 - right now, it's still bricked due to the DSA dts without any DSA bits in the build.

Ansuel commented:
> > @pkgadd can you by chance test this on your g10 just to exclude smem problems?
> 
> Will do, as soon as I've recovered the g10 - right now, it's still bricked due to the DSA dts without any DSA bits in the build.

sorry for the breakage :( hope we sort all these problem when you will be able to recover the device

pkgadd commented:
Don't worry, the problem is just that I'm physically distant from the device - and when I get home over a weekend, time might not suffice to work on it. Sadly the g10 cannot be recovered without serial console (it's good at noticing a non-booting kernel after flashing, by failing over to the OEM firmware, but if it boots up the first time (just without network), there's no way to recover it without serial).

As I've mailed you privately, there's sadly not a lot to see - no messages from kernel v5.15 at all, it just reboots into OEM.

Ansuel commented:
> Don't worry, the problem is just that I'm physically distant from the device - and when I get home over a weekend, time might not suffice to work on it. Sadly the g10 cannot be recovered without serial console (it's good at noticing a non-booting kernel after flashing, by failing over to the OEM firmware, but if it boots up the first time (just without network), there's no way to recover it without serial). So when it breaks, it is likely to remain broken for longer.
> 
> As I've mailed you privately, there's sadly not a lot to see - no messages from kernel v5.15 at all, it just reboots into OEM.

the fact that no error are printed it's strange... we have a way to know what is crashing before serial is available but don't worry robimarko is helping me as he also have a g10

koerhen commented:
Good news. The dts change made the difference: No bootloop any longer.

root@OpenWrt:~# uname -a
Linux OpenWrt 5.15.74 #0 SMP Sun Oct 23 21:33:22 2022 armv7l GNU/Linux
 
Note: I applied all patches of the pull request, so not just the dts change.
git status
On branch master
Your branch is up to date with 'origin/master'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	target/linux/generic/backport-5.15/301-v5.16-soc-qcom-smem-Support-reserved-memory-description.patch
	target/linux/generic/backport-5.15/407-v5.17-mtd-parsers-qcom-Don-t-print-error-message-on-EPROBE.patch
	target/linux/ipq806x/patches-5.15/130-mtd-nand-raw-qcom_nandc-handle-ret-from-parse-with-c.patch
	target/linux/ipq806x/patches-5.15/131-ARM-dts-qcom-ipq8064-disable-mmc-ddr-1_8v-for-sdcc1.patch

Ansuel commented:
Very nice! Now I just need to understand the g10 problem.

Il Lun 24 Ott 2022, 08:42 koerhen ***@***.***> ha scritto:

> Good news. The dts change made the difference: No bootloop any longer.
>
> ***@***.***:~# uname -a
> Linux OpenWrt 5.15.74 #0 SMP Sun Oct 23 21:33:22 2022 armv7l GNU/Linux
>
> Note: I applied all patches of the pull request, so not just the dts
> change.
> git status
> On branch master
> Your branch is up to date with 'origin/master'.
>
> Untracked files:
> (use "git add ..." to include in what will be committed)
>
> target/linux/generic/backport-5.15/301-v5.16-soc-qcom-smem-Support-reserved-memory-description.patch
>
> target/linux/generic/backport-5.15/407-v5.17-mtd-parsers-qcom-Don-t-print-error-message-on-EPROBE.patch
>
> target/linux/ipq806x/patches-5.15/130-mtd-nand-raw-qcom_nandc-handle-ret-from-parse-with-c.patch
>
> target/linux/ipq806x/patches-5.15/131-ARM-dts-qcom-ipq8064-disable-mmc-ddr-1_8v-for-sdcc1.patch
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/11032#issuecomment-1288494877>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQWAQ7Z5GFZBYHITYDLWEYVWJANCNFSM6AAAAAARKH5PJE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

robimarko commented:
Nice catch on the 1.8V advertisement, no idea how was that supposed to work if regulator that is passed is just a 3.3V fixed one.

Ansuel committed:
generic: 5.15: add missing CMDLINE_OVERRIDE patch

This patch was wrongly dropped with the assumption that it was moved to
generic. This wasn't the case and caused the malfunction of the Asrock
G10 router.

Reintroduce it to fix Asrock G10 functionality.

Fixes: 8cc2caed58e7 ("ipq806x: 5:15: add testing kernel version")
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel commented:
@koerhen can you provide a tested-by tag? It's optional but it would be good to have as i want to push this fix upstream in the kernel
`Tested-by: Name Surname <email>`

Ansuel commented:
@pkgadd also found the cause for g10 regression... it's the damn CONFIG_CMDLINE_OVERRIDE patch... there was an idea of making it generic but i totally missed the fact that nobody did that and here we have with broken devices and a patch that is the same for 4 different target and still not in generic...

pkgadd commented:
Great to hear, thanks for looking into that :)

hnyman commented:
FYI, There is apparently still something that can make R7800 unresponsive after a while.

I have now twice run into situation where R7800 does not respond to either wired or wlan, and also the reset pinhole does not work.

I mailed two weeks ago a short two-line serial log extract to @Ansuel, according to which CPU was somehow stuck , but this time I did not get any log (my bad, as I likely opened putty wrongly)

Note that the first lockup happened before the fixes from this PR, but the recent one already included this PR.

tapper82 commented:
My r7800 keeps rebooting on latest master.

Ansuel commented:
@tapper82 what do you mean with this? low uptime?

tapper82 commented:
Yes mate low uptime. EG yesterday I had 4 reboots through out the day. Then today it's bin up for 13h 21m 38s

hnyman commented:
@Ansuel 
My R7800 with r21123-641e4f2f04 had rebooted after almost 2 days uptime with the following pstore crash file.

I noticed this just now, over a day since the reboot. Curiously, the router has rebooted at an almost idle moment (when I have been away).

Looks like a CPU clock rate / cpufgreq thing. Normal ondemand in use.
The `virtual address ffffff0c` looks like a pointer error with struct, but a negative offset??? 
Untypical. 
An error code?

```
<6>[  120.837140] br-lan: port 2(phy0-ap0) entered blocking state
<6>[  120.842454] br-lan: port 2(phy0-ap0) entered forwarding state
<4>[25849.620485] ath10k_pci 0000:01:00.0: Invalid VHT mcs 15 peer stats
<1>[159699.472834] 8<--- cut here ---
<1>[159699.472870] Unable to handle kernel paging request at virtual address ffffff0c
<1>[159699.474806] pgd = 69b372ab
<1>[159699.482186] [ffffff0c] *pgd=5fdfd861, *pte=00000000, *ppte=00000000
<0>[159699.484862] Internal error: Oops: 37 [#1] SMP ARM
<4>[159699.491455] Modules linked in: pppoe ppp_async ath10k_pci ath10k_core ath wireguard pptp pppox ppp_mppe ppp_generic nft_redir nft_nat nft_masq nft_flow_offload nft_fib_inet nft_ct nft_chain_nat nf_nat                                  nf_flow_table_ipv6 nf_flow_table_ipv4 nf_flow_table_inet nf_flow_table nf_conntrack_netlink nf_conntrack mac80211 libchacha20poly1305 ipt_REJECT curve25519_neon cfg80211 xt_time xt_tcpudp xt_tcpmss xt_stati                                 stic xt_multiport xt_mark xt_mac xt_limit xt_length xt_hl xt_esp xt_ecn xt_dscp xt_comment xt_TCPMSS xt_LOG xt_HL xt_DSCP xt_CLASSIFY slhc sch_cake poly1305_arm nft_reject_ipv6 nft_reject_ipv4 nft_reject_ine                                 t nft_reject nft_quota nft_objref nft_numgen nft_log nft_limit nft_hash nft_fib_ipv6 nft_fib_ipv4 nft_fib nft_counter nft_compat nf_tables nf_reject_ipv4 nf_log_syslog nf_defrag_ipv6 nf_defrag_ipv4 libcurve2                                 5519_generic libcrc32c iptable_mangle iptable_filter ipt_ah ipt_ECN ip_tables crc_ccitt compat chaoskey chacha_neon fuse cls_bpf act_bpf sch_tbf sch_ingress sch_htb sch_hfsc em_u32
<4>[159699.492239]  cls_u32 cls_tcindex cls_route cls_matchall cls_fw cls_flow cls_basic act_skbedit act_mirred act_gact ledtrig_usbport xt_set ip_set_list_set ip_set_hash_netportnet ip_set_hash_netport ip_s                                 et_hash_netnet ip_set_hash_netiface ip_set_hash_net ip_set_hash_mac ip_set_hash_ipportnet ip_set_hash_ipportip ip_set_hash_ipport ip_set_hash_ipmark ip_set_hash_ipmac ip_set_hash_ip ip_set_bitmap_port ip_set                                 _bitmap_ipmac ip_set_bitmap_ip ip_set nfnetlink ip6table_mangle ip6table_filter ip6_tables ip6t_REJECT x_tables nf_reject_ipv6 msdos ip_gre gre ifb ip6_udp_tunnel udp_tunnel sit tunnel4 ip_tunnel tun hfsplus                                  cdrom cifs oid_registry cifs_md4 cifs_arc4 asn1_decoder dns_resolver nls_utf8 nls_iso8859_15 nls_iso8859_1 nls_cp850 nls_cp437 nls_cp1250 sha1_generic seqiv md5 ecb des_generic libdes cmac usb_storage leds_                                 gpio xhci_plat_hcd xhci_pci xhci_hcd dwc3 dwc3_qcom ohci_platform ohci_hcd phy_qcom_ipq806x_usb ahci fsl_mph_dr_of ehci_platform ehci_fsl sd_mod ahci_platform libahci_platform
<4>[159699.561952]  libahci libata scsi_mod scsi_common ehci_hcd ramoops reed_solomon pstore gpio_button_hotplug vfat fat ext4 mbcache jbd2 exfat crc32c_generic
<4>[159699.670552] CPU: 1 PID: 31923 Comm: kworker/1:2 Not tainted 5.15.74 #0
<4>[159699.684252] Hardware name: Generic DT based system
<4>[159699.690761] Workqueue: events dbs_work_handler
<4>[159699.695621] PC is at __clk_notify+0x4c/0x94
<4>[159699.700133] LR is at srcu_notifier_call_chain+0x9c/0xf4
<4>[159699.704649] pc : [<c06a4238>]    lr : [<c03438e8>]    psr: a0000013
<4>[159699.709944] sp : c73a9da8  ip : 60000013  fp : c1769f80
<4>[159699.716281] r10: 000927c0  r9 : 23c34600  r8 : 66d16940
<4>[159699.721576] r7 : 00000002  r6 : c1656b40  r5 : c0d8086c  r4 : ffffff0c
<4>[159699.726872] r3 : c17ff9c0  r2 : c1656840  r1 : 00000002  r0 : 00000001
<4>[159699.733210] Flags: NzCv  IRQs on  FIQs on  Mode SVC_32  ISA ARM  Segment none
<4>[159699.739812] Control: 10c5787d  Table: 44b9006a  DAC: 00000051
<1>[159699.747099] Register r0 information: non-paged memory
<1>[159699.752912] Register r1 information: non-paged memory
<1>[159699.758034] Register r2 information: slab kmalloc-192 start c1656840 pointer offset 0 size 192
<1>[159699.763160] Register r3 information: slab kmalloc-64 start c17ff9c0 pointer offset 0 size 64
<1>[159699.771755] Register r4 information: non-paged memory
<1>[159699.780428] Register r5 information: non-slab/vmalloc memory
<1>[159699.785466] Register r6 information: slab kmalloc-192 start c1656b40 pointer offset 0 size 192
<1>[159699.791286] Register r7 information: non-paged memory
<1>[159699.799787] Register r8 information: non-paged memory
<1>[159699.804995] Register r9 information: non-paged memory
<1>[159699.810116] Register r10 information: non-paged memory
<1>[159699.815240] Register r11 information: slab kmalloc-128 start c1769f80 pointer offset 0 size 128
<1>[159699.820364] Register r12 information: non-paged memory
<0>[159699.829388] Process kworker/1:2 (pid: 31923, stack limit = 0x8eb59e4d)
<0>[159699.834339] Stack: (0xc73a9da8 to 0xc73aa000)
<0>[159699.840938] 9da0:                   c158be00 c17ffec0 66d16940 23c34600 c158be00 23c34600
<0>[159699.845458] 9dc0: c1656b40 00000000 c15a80c0 c06a8880 c1656a28 c15a80c0 c06b5d38 c140b180
<0>[159699.853704] 9de0: 66d16940 23c34600 000927c0 c06a88c4 c1656b40 00000000 23c34600 c15a80c0
<0>[159699.861952] 9e00: dd99a010 23c34600 000927c0 c06a8af0 23c34600 23c34600 00000000 ffffffff
<0>[159699.870198] 9e20: 23c34600 c0d82a58 c73a9ef8 c176a640 23c34600 00000001 c1f4e100 dd99a010
<0>[159699.878445] 9e40: 23c34600 c06a8c70 c176ce00 c15ae180 00000001 c1f4e100 dd99a010 c07ddbbc
<0>[159699.886690] 9e60: c1656b40 c1f4e100 c73a9ec8 23c34600 00000000 c176ce00 dd99a010 00000006
<0>[159699.894938] 9e80: 23c34600 c15ae180 00000000 000927c0 00000000 c07ddf6c c0d90408 c03438c8
<0>[159699.903183] 9ea0: c176d800 23c34600 c176d800 c0dd0004 00000000 00000001 001a5248 c07e2784
<0>[159699.911430] 9ec0: c176d800 001a5248 000927c0 00000024 00000000 c176d800 c1f4e300 c1f4e280
<0>[159699.919677] 9ee0: c1f4e300 c176a7c0 c1f4e280 dd9a0305 c4223540 c07e59ec c1f4e338 00000000
<0>[159699.927925] 9f00: c1f4e304 c0d906ec 00000000 00000040 dd9a0305 c07e66c8 c1f4e338 c4223500
<0>[159699.936169] 9f20: dd99d040 dd9a0300 00000000 c0339fb8 c73a8000 dd99d040 00000008 c4223500
<0>[159699.944417] 9f40: c4223518 dd99d040 00000008 dd99d058 c0d03d00 dd99d200 c73a8000 c033a2b8
<0>[159699.952663] 9f60: c0d0c36c c0d9aadc c3b3fecc c5f42200 c4379300 c033a244 c4223500 c73a8000
<0>[159699.960909] 9f80: c3b3fecc c4379320 00000000 c0341f84 c5f42200 c0341e2c 00000000 00000000
<0>[159699.969155] 9fa0: 00000000 00000000 00000000 c0300130 00000000 00000000 00000000 00000000
<0>[159699.977407] 9fc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
<0>[159699.985648] 9fe0: 00000000 00000000 00000000 00000000 00000013 00000000 00000000 00000000
<0>[159699.993894] [<c06a4238>] (__clk_notify) from [<c06a8880>] (clk_change_rate+0xfc/0x2b8)
<0>[159700.002138] [<c06a8880>] (clk_change_rate) from [<c06a88c4>] (clk_change_rate+0x140/0x2b8)
<0>[159700.010037] [<c06a88c4>] (clk_change_rate) from [<c06a8af0>] (clk_core_set_rate_nolock+0xb4/0x1f8)
<0>[159700.018371] [<c06a8af0>] (clk_core_set_rate_nolock) from [<c06a8c70>] (clk_set_rate+0x3c/0x170)
<0>[159700.027400] [<c06a8c70>] (clk_set_rate) from [<c07ddbbc>] (_set_opp+0x250/0x518)
<0>[159700.036427] [<c07ddbbc>] (_set_opp) from [<c07ddf6c>] (dev_pm_opp_set_rate+0xe8/0x218)
<0>[159700.043803] [<c07ddf6c>] (dev_pm_opp_set_rate) from [<c07e2784>] (__cpufreq_driver_target+0x194/0x238)
<0>[159700.051619] [<c07e2784>] (__cpufreq_driver_target) from [<c07e59ec>] (od_dbs_update+0xcc/0x1a0)
<0>[159700.060995] [<c07e59ec>] (od_dbs_update) from [<c07e66c8>] (dbs_work_handler+0x38/0x74)
<0>[159700.070019] [<c07e66c8>] (dbs_work_handler) from [<c0339fb8>] (process_one_work+0x230/0x4bc)
<0>[159700.078096] [<c0339fb8>] (process_one_work) from [<c033a2b8>] (worker_thread+0x74/0x5d4)
<0>[159700.086511] [<c033a2b8>] (worker_thread) from [<c0341f84>] (kthread+0x158/0x174)
<0>[159700.094672] [<c0341f84>] (kthread) from [<c0300130>] (ret_from_fork+0x14/0x24)
<0>[159700.102135] Exception stack(0xc73a9fb0 to 0xc73a9ff8)
<0>[159700.109253] 9fa0:                                     00000000 00000000 00000000 00000000
<0>[159700.114467] 9fc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
<0>[159700.122712] 9fe0: 00000000 00000000 00000000 00000000 00000013 00000000
<0>[159700.130962] Code: e59440f4 e1550004 e24440f4 0a00000a (e5943000)
<4>[159700.137933] ---[ end trace 936bddb7763ad5ce ]---
```

Ansuel commented:
@hnyman the problem here and on all the other crash is cache corruption....

on the original fw the clk is correctly set to 1ghz or 1.2ghz but on our fw it's 200mhz less and changes a lot hinting there is something wrong in how we setup that clk. I'm messing by building the original r7800 from netgear gpl source hoping to repro our bug in the original fw and finding what we are missing. This is why I'm not implementing my idea of a workaround for this cache crash... (probably the new devfreq driver is too fast and the problem is more present)
